explorer/service/src/main/java/org/radixdlt/explorer/metrics/MetricsProvider.java

```diff
@@ -178,25 +170,62 @@ private void calculateMetrics(Map<String, SystemInfo> info) {
         long progress = Math.round(aggregatedProgress / nodeCount);
 
         synchronized (calculationLock) {
-            calculationIterations++;
             long seconds = (System.currentTimeMillis() - testState.getStartTimestamp()) / 1000 + 1;
             assert seconds > 0;
             averageTps = Math.round(progress / seconds);
             peakTps = Math.max(peakTps, tps);
             calculatedMetrics = new Metrics(tps, progress, averageTps, peakTps);
+            dumpCurrentMetrics();
```

Comment:
Not good idea to do I/O while holding a lock (hint: I/O operations take milliseconds to complete)

I think an asynchronous (periodically scheduled) task is a better design for dumpingCurrentMetrics.
A caveat is to only execute the Task if the test is in RUNNING state.

Reply:
I fear moving the "dump-to-file" operation outside the synchronized context may force us into magical race conditions where we're not writing what we just calculated (this method may be called at any time, from any thread).

Any further input on how to prevent these race conditions from a separate worker thread would be greatly appreciated.

Reply:
If you do all the writing from a single background thread then there is no need to synchronize dumpingCurrentMetrics() ?

Reply:
Ah! But wait! You, of course, mean that I should "enqueue" a new "dump task" and let that queue be serialized from a single background thread.

Now it makes sense! Give me a coupe of minutes to fix this! :)

Reply:
Now fixed!

Reply:
Sorry for being unclear. Yes, you got my point correctly you serialised writes to the file and thus prevented potential race conditions.